qiskit/backends/basebackend.py

```diff
@@ -69,10 +69,10 @@ def status(self):
         backend_name = self.configuration.get('name', '')
         return {'name': backend_name, 'available': True}
 
-    def __str__(self):
+    def __repr__(self):
```

Comment:
Actually I toyed with the idea of using `__repr__` during the initial PR, but in the end went with `__str__` as they have different purposes. From https://docs.python.org/3/reference/datamodel.html#object.__repr__:
>  If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned.

While for `__str__`:
> This method differs from object.__repr__() in that there is no expectation that __str__() return a valid Python expression: a more convenient or concise representation can be used.

Ie. ideally, the string returned by `__repr__` should be enough to build a backend based on it - which I think is not a trivial problem to solve (in particular, for the remote backends, you cannot only use the backend name - you need information about the provider (or the api used by the provider) in order to recreate the same backend fully). It's true that `__str__` does not get called automatically when printing a list, but I feel that can be considered "by design", and actually avoids some confusion (ie. I think it's good that we provide "handy" printable strings, but the user should know he is using objects).

And there are some other implications for `repr` that are no so evident (mostly for debugging and inspection) - bottom line, I think it is safer to keep using `__str__` instead of `__repr__`, at least during this iteration - would it make sense?

Reply:
Having just __str__ would sort of defeat the purpose of an easy way to filter backends and see them. In a way, a filter *is* a convenience we provide. Otherwise users can just get all the backends and filter on their own.

I personally find it a bit weird that Python does not implement list iteration for `__str__`; it might change in the future. A `print` statement in Python usually works on lists too. Meanwhile, using `__str__` means we are forcing the user to think about this Python implementation detail. I don't agree that it forces them to think object vs. string (otherwise why even implement `__str__`?)... it is just confusing behavior: when the result of filter is one object, it prints fine. When it is a list of backends, you have to write a for loop to print it.

While this useability is my main reason for using `__repr__`, I can also argue that it *is* possible to unambiguously reconstruct a backend from its string, except for the extremely rare case of 2 separate providers using the same backend name. i.e you would just call `get_backend()` on all providers, and use the object that returned anything other than None.

Reply:
Python weirdness aside (yeah, it's not intuitive that `print(backend)` invokes str() and `print([backend])` does not), I really don't think directly contravening the Python documentation for a standard object of a method is something we should do, specially "only" for a cosmetic gain. Again, `__repr__` should return:

> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned.

Which I don't think we can achieve in general (it's not only in extremely rare cases - for fully recreating a remote backend, we need information about the provider and the authentication as well, and feels like entering muddy waters). If we are not 100% sure that we can, I'd really encourage not to use `__repr__` (and if we have to also remove `__str__` in the process to avoid confusion, so be it!).

Reply:
Applying `__str__` on list elements seems to have been proposed in Python, but got struck down by the Python dictator himself:
https://www.python.org/dev/peps/pep-3140/#rejection

No good reason given, except that "we're too close to a beta release." This was in 2008.